pkg/label/label.go

```
@@ -33,28 +33,54 @@ func ComputeLabels(u *unstructured.Unstructured) error {
 	} else if found {
 		newLabels = map[string]string{}
 	} else {
-		newLabels = removeLabelsWithKRMPrefix(u.GetLabels())
+		newLabels = u.GetLabels()
 	}
+	// No matter where the labels come from, sanitize them based on GCP label validation.
+	newLabels = SanitizeGCPLabels(newLabels)
 	newLabels[CnrmManagedKey] = "true"
 	return unstructured.SetNestedStringMap(u.Object, newLabels, "spec", "labels")
 }
 
 func NewGCPLabelsFromK8sLabels(labels map[string]string) map[string]string {
-	res := removeLabelsWithKRMPrefix(labels)
+	res := SanitizeGCPLabels(labels)
 	// Apply default label.
 	res[CnrmManagedKey] = "true"
 	return res
 }
 
-func removeLabelsWithKRMPrefix(labels map[string]string) map[string]string {
+// Sanitize labels with GCP label validation
+func SanitizeGCPLabels(labels map[string]string) map[string]string {
 	res := make(map[string]string)
+keyLoop:
 	for k, v := range labels {
-		if len(strings.Split(k, "/")) == 2 {
-			// Do not include any KRM-style labels (labels that include a prefix
-			// denoted with a '/').
-			// TODO(b/137755194): Determine long-term solution.
+		// GCP labels have strict validation rules. This function filters out any
+		// labels from Kubernetes metadata that do not conform to these rules,
+		// preventing them from being propagated to GCP.
+		// See: https://cloud.google.com/compute/docs/labeling-resources#requirements
```

I've been considering this, and my personal view is that we should not filter all invalid labels. If an invalid label is added by customer, the controller should still pass it to the server, allowing it to fail with an error instead of silently ignoring it. Our sanitization should be limited to KRM-style labels that are automatically added by Config Sync or by us and are known to fail GCP's label validation. But I'm open to either approach.